### DISCUSSION TITLE
Add dwim push that uses upstream vs pushRemote or block upstream push if pushRemote exists

### DISCUSSION THREAD

**Discussion by darkfeline** [magit/magit#4707] [Ideas]
Currently Magit requires the user to explicitly choose whether to push to upstream (`P u`) or pushRemote (`P p`).  There should be a single dwim command that intelligently picks between the two like how `git push` does it.

The reason this is bad is because the user may have muscle memory from pushing a branch that has no pushRemote, then accidentally pushing to upstream on a branch with pushRemote set, where pushing to upstream is a bad idea (like pushing a WIP feature branch to master).

Alternatively, there should be some way to prevent the user from accidentally pushing to upstream when pushRemote is set, like requiring a yes/no prompt.

Example for the latter: https://github.com/darkfeline/magit/commit/1e27996767a933a264140f8f0d83a38593602913

**Comment by kyleam** [comment]
> Alternatively, there should be some way to prevent the user from accidentally pushing to upstream when pushRemote is set, like requiring a yes/no prompt.

I like that idea, and on my end I'd probably enable confirmations (though for topic branches, I typically configure my upstream to a local branch, so there's no real risk of an accidental push).

> There should be a single dwim command that intelligently picks between the two like how git push does

`magit-push-to-remote` and `magit-push-implicitly` call `git push` without a refspec.  Perhaps you want to add one of those to your push transient?

**Reply by darkfeline** [reply]
`magit-push-to-remote` takes a remote so that doesn't work.  `magit-push-implicitly` looks like the first feature request here, so I have added it to my config.  I still want the push confirmation to prevent accidents.

**Reply by tarsius** [reply]
> Alternatively, there should be some way to prevent the user from accidentally pushing to upstream when pushRemote is set, like requiring a yes/no prompt.

How about this approach. If `p p` pushes to the same remote branch as `p u`, then doing it with `p p` might indicate that the user wasn't aware what they were doing. We could add an option to allow configuring to either ask or just bail (and suggest to use `p u` instead).

**Reply by darkfeline** [reply]
This is more about when `p p` doesn't push to the same remote branch as `p u`, and the user does `p u` accidentally from muscle memory.

**Reply by tarsius** [reply]
It surprises me that `p u` is the problematic binding.

To me `p p` is the "let others see my work in progress" command (which I use it a lot, so I might end up using it as a reflex) and `p u` is the "this is done now, let's make it official" command (which I only use consciously).

Please read [the two remotes](https://magit.vc/manual/magit/The-Two-Remotes.html) and let me know your position on the matter.

**Reply by MatthewFluet** [reply]
I've occasionally done an accidental `P u` when I should have done a `P p`.  It tends to happen when I'm dividing my time between projects, one of which is collaborative (and the two remotes is appropriate, so I should do `P p` followed by a pull request) and another is semi-private (commits are made directly to `main`/`master` and then pushed to `origin` with `P u`).  In the latter case, I should probably explicitly set `pushRemote` to `origin` and always do `P p`.  Though, that would work against @tarsius's suggestion:

> How about this approach. If `p p` pushes to the same remote branch as `p u`, then doing it with `p p` might indicate that the user wasn't aware what they were doing. We could add an option to allow configuring to either ask or just bail (and suggest to use `p u` instead).

**Reply by darkfeline** [reply]
> It surprises me that `p u` is the problematic binding.

Well, `p u` is the dangerous one.  If you set a pushRemote, then you generally plan to push to it.  But in a triangular workflow where you pull from upstream and push to pushRemote, then yes, `p u` is the dangerous command.

> 
> To me `p p` is the "let others see my work in progress" command (which I use it a lot, so I might end up using it as a reflex) and `p u` is the "this is done now, let's make it official" command (which I only use consciously).

Then you do not have a lot of private/personal repos where you push and pull from a single remote master branch.  I do, and I think many people do considering Git's default behavior, which is to set upstream and not pushRemote. The Git docs even call out centralized workflows in [push.default](https://git-scm.com/docs/git-config#Documentation/git-config.txt-pushdefault).  We use `p u` to push in centralized repos and `p p` to push in feature branch repos.  Hopefully you can see why that would be a problem.

> Please read [the two remotes](https://magit.vc/manual/magit/The-Two-Remotes.html) and let me know your position on the matter.

My position on the matter is that not all repos do exactly what you do.  I don't think there's anything wrong; heck, I think it's great.  I even use it for some repos.  But not all repos, for various reasons.  That's what this thread is about.

**Reply by tarsius** [reply]
You clearly understand the triangular workflow and other aspects of all of this, which makes it easier to have a conversation about it, compared to other conversations I was forced to have about it in the past. That being said, those past conversations have worn me out, and I don't have the mental energy to do it all over again. (Which might also have to do with the fact that it is 38C in my room right now, at 21:00.)

The "two remotes" section was an earlier attempt at making these conversations unnecessary, but I can see that that is insufficient. I will have to write another article about it, which not only explains what it is and why you would want to use it in certain cases, but also (1) how to set it up and (2, and this is the conversation I would have to have with you) why it makes sense to setup *all* repositories for the triangular workflow, even the repositories in which one does not take advantage of that setup.

I also realize that (1) I will never succeed in convincing everyone that this is the way, (2) there are indeed exceptions where it doesn't make sense (3) I cannot, and do not desire to, force every magit user to do it my way and (4) that means that I have to add features that I think are unnecessary or even misguided. (I do, however, reserve the right to add tongue in cheek documentation for those features :yum:.)

As I said, I have had conversations about this many times, each time focusing on a certain aspect, depending on what the disagreement and/or misunderstanding was focused on. The time has once more come where I need to try to explain my reasoning in full. This will take a while, I am not motivated to do it right now, I need to finish other things, and I want and need to take a break.

In short, I want to delay this conversation indefinitely. Not forever, but I certainly cannot promise that I will pick it up in, say, one month.

**Reply by darkfeline** [reply]
I don't see anything particularly objectionable in your reply, but I'm not sure how it argues against either idea being proposed.

The dwim command idea (which looks like would just be binding `magit-push-implicitly` by default) does not affect people exclusively using triangular flows like yourself, nor people exclusively using centralized flows.  It improves the experience of users using both flows.  So it's strictly an improvement (unless you want to wage a philosophical war with people that are using flows you think are inferior).

The dwim would likely be bound to `p i`, and `p u` and `p p` would continue to exist, so it also doesn't affect people who would prefer using the existing commands.

The "add a warning to `p u` if pushRemote is set and different` idea seems compatible with your workflow:

> `p u` is the "this is done now, let's make it official" command (which I only use consciously)

You get an explicit check before doing the intentional command.  Maybe you find typing that extra `y` or `yes` to be inconvenient; of course such a check would likely be configurable for people who want to live on the edge (although I think you would want it enabled in case something falls on your keyboard).

Addendum:

Actually, there's some bikeshedding that can be done for the dwim idea.  I think a case could be made for making the dwim command the **only** default action (in addition to the `p e` push elsewhere action).  In a "standard" triangular workflow, you never push to upstream; instead, upstream merges directly from your pushRemote:

```
fred     wilma
  *<------*
   \     ^
    \   /
     v /
      *
     you
```

(from https://stackoverflow.com/a/63144528/469721)

For example, on GitHub, the pushRemote is associated with a PR, and the upstream merge happens by the upstream owner merging the PR, rather than you pushing to upstream.  Even if you as the PR owner have commit rights to upstream, I think you would want to pull the PR from upstream, rather than push to upstream locally (e.g., you would merge the feature branch into master and then push master to upstream origin/master, rather than push the feature branch directly to upstream origin/master)

So there's never a need to explicitly choose to push to either upstream or pushRemote.  Which is to say, your workflow sounds unusual.  Of course, you're entitled to your workflow and I cannot say that it's wrong.

**Reply by tarsius** [reply]
> I don't see anything particularly objectionable in your reply, but I'm not sure how it argues against either idea being proposed.

I am very much in a hurry and pretty much out the door, so I could not read further than that. There is *nothing* objectionable in your reply and I didn't try to imply otherwise. I just wanted to say, that I don't really want to have a conversation about this right now. I actually thought about it a lot the last two days and realized I need to make some improvements. I don't know exactly what they will be yet, but I have to think about them. Of course thinking about them could, and probably will, include talking to others about it, but I am not quite there yet.

I could focus on getting this resolved quickly but the motivation for that would be to get it done with, not to arrive at the best possible result. So the point I was trying to get across was this: please give me some time, this will lead to a better result and also that this isn't my top priority right now and I don't want to make it my top priority just because it happens that someone brought it up right now.

**Reply by tarsius** [reply]
ah nothing objectonable in what *i* said. 🤪

**Reply by tarsius** [reply]
(I've been using `p ...` because I forgot that that is a personal customization.)

> Then you do not have a lot of private/personal repos where you push and pull from a single remote master branch.

The majority of my repositories are like that. Your assumption that I have come up with a setup that works very well for one particular workflow, but not so well for others, is wrong. I use the same basic setup in repositories that use very different workflows and it works equally well for all of them.

> I think a case could be made for making the dwim command the only default action

Exactly. I think we are actually very much in agreement on how it should work. We disagree on how that goal should be achieved.

- You want to see it implemented using either the "dwim" command `git push` (without arguments) or in a way similar to that. And the option you want to use to customize the behavior in cases where that is necessary is `push.default`.

  I think plain old `git push` is a fine command for us *in a terminal* -- but it is completely unsuitable for a git tui/gui client to base its own dwim command on, exactly because `push.default` (and several other variables) can be used to make it behave in vastly different ways.

- My approach is based on the `remote.pushDefault` variable. This variable was a later addition and it does not fall back to a reasonable default value when unset, which I believe are the reasons why it isn't being embraced more.

   This variable is not only suitable for repositories that use a triangular workflow, but I do realize that I contribute to that misconception by introducing it in magit's documentation in that context. And granted, it might feel a bit weird to set the "default push remote" in a repository that has only a single remote.

Here are some common setups that I use:

1. The single branch setup (i.e., `main` and its remote copy `origin/main`).

   This is how all my repositories start out, and a great many of them remain in that state.

   Initial setup:

   ```
   git init
   git remote add origin origin git@mine
   git push --set-upstream origin main
   git config remote.pushDefault origin
   ```

   Going forward I use `p p` to push `main` to `origin/main`.

2. Feature branches in my own repository.

   The initial setup is the same as above and no additional setup is required. I just create a new branch `new-feature` and push it to `origin/new-feature` using `p p`, the same key binding as I used to push `main`.

3. Contribute to someone else's repository.

   Initial setup:

   ```
   git clone git@theirs
   git remote add mine git@mine
   git config remote.pushDefault mine
   ```

   Now there are two remotes. `remote.pushDefault` is always set to the remote I can actually push to, which in this case is `mine`. Above it was `origin`, which also happens to be the only remote in those case.

   After implement a new feature on branch `new-feature`, I push it to `mine/new-feature` using `p p`. Sometimes I make changes on `main`, either because I don't want to contribute them upstream or because I want to immediately use changes from multiple feature branches without having to wait for upstream. In such cases I make them on `main`, which I then push to `mine/main` with `p p`.

4. Someone contributes to my repository.

   Initial setup was described in (2).

   I use my Forge package to inspect, possibly modify, and merge pull-requests. This begins by pressing `b n` on the pull-request shown in the magit status buffer. This adds a remote for the user's repository, fetches the appropriate branch, sets the upstream, and sets`branch.contributed-feature.pushRemote` to `somebody`.

   If I do not need to make changes to the feature branch I then press `m i RET`, which merges the feature branch into `main`, it also deletes my local copy of the feature branch and the remote from which it came (unless I have checked out another pull-request/feature branch from the same contributor). Then I push `main` to `origin/main` using `p p`.

   Sometimes I make trivial changes to feature branches authored by a contributor. After doing so, I push them to `somebody/contributed-feature` using `p p`. Then I proceed to merging as described before.

Note

1. All pushing was done using `p p`. It dwims but there is one requirement: `remote.pushDefault` has to be set. Magit helps with setting that variable but there are some situation where it doesn't get set automatically (I should improve that).
2. The remote branch that the current branch is being pushed to always has the same name as the local branch. (In the rare cases when you want to push to remote branch with another name, you would have to use another command such as `p e`. But as I understand it, like me, you consider that a special case and are okay with having to use another key binding in that case.)
3. It makes a lot of sense to use the variable `remote.pushDefault` to configure where you want to push -- I mean, the words that its name is made up from are also contained in the sentence "the remote where I want to push to by default". It is not only suitable for triangular workflows. Sometimes you want to push to another remote than the default -- for pull-requests the forge package takes care of setting this up automatically.

**Reply by kyleam** [reply]
@tarsius I'm mostly staying out of this discussion at this point, but, given your workflow description, I want to clarify my initial response: the workflows you describe very closely align to mine.  In all of those different scenarios, I too use `magit-push-current-to-pushremote`.  Yet I still replied that I like the idea of adding (an optional) confirmation for `magit-push-current-to-upstream` when the push remote is configured.

It is because I have a setup where I nearly always push to configured the push remote that I want to guard against making a dumb mistake of pushing somewhere I didn't intend.  As I mentioned in the original post, I currently side step this by setting the upstream of topic branches to a local branch (e.g., master or maint).  However, for some repos, I would rather use origin/master or origin/maint, but I avoid doing so because I have write-access to origin and I don't want to risk an accidental push.  (Obviously I don't find using local upstreams too inconvenient given I haven't tweaked my configuration to require confirmation.)

If I recall correctly, the above approach (using a local upstream) has only saved me from one accidental push, but I'm very glad it did.

**Reply by darkfeline** [reply]
I find it hard to follow you.  Are you suggesting a particular new UI or are you saying the current UI is fine?

The point I and kyleam are making is that the current UI makes it easy to do a dangerous push for many average users (this isn't about your workflow tarsius, I understand you have an awesome workflow that doesn't have this issue).  Users with simple centralized checkouts push with `P u`.  Users with triangle workflows push with `P p`.  Users with both will occasionally do a `P u` to disastrous effect, and the current UI does not prevent or warn against this.

**Reply by tarsius** [reply]
I don't oppose adding confirmation dialogues.

I have pushed a stub implementation to the `confirm-push` branch. The final implementation does not have to be based on that, but it should take the concerns/special-cases that I raise there into account. I find myself unable to push through and implement this myself at this point, so someone else will have to do the work.

It has been coming for a while, but I just went over the edge and have burned out. Effective immediately I am taking a mental health break lasting at least a month.

Kyle, sorry for dropping this on you like this. Please add the requested confirmation dialogues as you see fit and if you conclude that #4715 is an improvement, then please merge it. Likewise for other pull-requests that come up while I am gone.

**Reply by tarsius** [reply]
(Also look at ce5d2b3cd34fc25d06bd1f6cb6beef0703d076ed from the `idea` branch. Maybe that is useful, maybe not.)

**Reply by kyleam** [reply]
> It has been coming for a while, but I just went over the edge and have burned out.

Sorry to hear that.  I hope the break gets you to a better (and long-lasting) state.

**Reply by tarsius** [reply]
It turned out it wasn't burnout but Covid.

Anyway, I am feeling burnout endangered again, so not picking up here any time soon.

Mainly here to say "once I actually work on this, I should also consider the suggestion in #5482".
